File: Python/222. Count Complete Tree Nodes.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def countNodes(self, root: TreeNode) -> int:
        def compute_depth(node):
            d = 0
            while node.left:
                node = node.left
                d += 1
            return d
        
        def exists(idx, d, node):
            left, right = 0, 2**d - 1
            for _ in range(d):
                pivot = left + (right - left) // 2
                if idx <= pivot:
                    node = node.left
                    right = pivot
                else:
                    node = node.right
                    left = pivot + 1
            return node is not None
        
        if not root:
            return 0
        d = compute_depth(root)
        if d == 0:
            return 1
        left, right = 1, 2**d -1
        while left <= right:
            pivot = left + (right - left) // 2
            if exists(pivot, d, root):
                left = pivot + 1
            else:
                right = pivot - 1
        return (2**d - 1) + left

File: 2023/222. Count Complete Tree Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
   bool exist(TreeNode* root, int depth, int target) {
        int left = 0, right = (int)pow(2, depth - 1) - 1;
        for (int i = 0; i < depth - 1; i++) {
            int mid = left + (right - left) / 2;
            if (mid < target) {
                root = root->right;
                left = mid + 1;
            } else {
                root = root->left;
                right = mid - 1;
            }
        }
        return root != NULL;
    }
public:
    int countNodes(TreeNode* root) {
         if (root == NULL) return 0;
        int depth = 0;
        TreeNode* node = root;
        while(node != 0) {
            depth++;
            node = node->left;
        }
        int left = 0, right = (int)pow(2, depth - 1) - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (exist(root, depth, mid)) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return (int)pow(2, depth - 1) - 1 + left;
    }
};